### PR TITLE
Batch LP certificate dual reads

### LINKED ISSUE #211: Explore vectorized dual retrieval in certified_lp_bound
Each LP certificate currently issues one `JuMP.dual` call per affine constraint per neuron bound solve. MOI can fetch `ConstraintDual` for a vector of constraint indices in a single call, which would cut per-read overhead through the JuMP/MOI/caching layers.

Profile before building: certification added 29.82% wall time in the 500-sample WK17a run, and #209's progressive tightening removes most unnecessary certificate work, so this only pays off if the dual reads (rather than the LP solves) dominate what remains.

Design note: a batched read fails as a unit, so the per-constraint error isolation in `solver_attribute_or_nothing` needs a fallback story (e.g. retry per-constraint on batch failure).

### PR DESCRIPTION
**Pull request by vtjeng**
## Why

LP certificate construction reads one dual value per affine constraint for every neuron-bound
solve. Verification models have thousands of rows, so the scalar path repeats JuMP result checks
and cache index mapping even though MathOptInterface can read a homogeneous vector of constraint
duals at once.

This overhead still matters after #209's progressive tightening. On the post-#209 500-input WK17a
baseline, non-solver bound overhead took 299.0 s, compared with 170.6 s inside HiGHS.

## What changes

- Read `ConstraintDual` once per affine function/set group through the vector MathOptInterface API.
- Expose a single injection seam: a `dual_values` callback mapping a constraint vector to duals.
  When a group read is unavailable, throws, or returns the wrong shape, retry each constraint
  through the same callback as a single-element batch, so a custom source stays authoritative at
  both granularities. (The earlier `dual_value` scalar seam and its interaction rule are gone.)
- Ignore invalid batch elements independently while continuing to propagate interrupts and
  resource-exhaustion errors.
- Process typed constraint and dual vectors behind a function barrier so the certificate loop stays
  specialized.
- Record the measured impact in PERFORMANCE.md.

Backends may implement the vector getter with scalar solver reads. The change still avoids repeated
JuMP and caching-layer work.

## Measured impact

Paired post-#209 master `8a455e2` against PR commit `5b40704` with
`--samples 1:500 --tightening lp --main-time-limit 120 --norm-order Inf`:

- Formulation: 866.2 s → 702.6 s, saving 163.6 s (18.9%). A paired input bootstrap gives a
  17.6–20.0% interval; this reflects input mix, not run-to-run machine variation.
- Non-solver bound overhead: 299.0 s → 183.5 s, saving 115.5 s (38.6%).
- Bound-solver wall time: 170.6 s → 170.2 s, with 99,067 solver calls on both sides.
- Summed end to end: 2,431.7 s → 2,270.7 s, saving 161.0 s (6.6%). Final-solve wall time was flat.
- The median build-and-tightening ratio was 0.81, and 93% of modeled inputs improved by more than
  1%.

The benchmarked commit is `5b40704`; later commits on this branch refactor the dual-read seams,
tests, and documentation without changing the successful-group-read hot path, and were not
re-benchmarked.

The 120 s final-solve limit produced expected solver-path churn: two baseline-unresolved inputs
became resolved, no semantic outcome regressed, and a same-commit control showed substantial
fresh-process solve-time variation. The formulation result is the useful attribution: model counts,
ReLU counts, bound calls, and bound status/skip counts matched.

[Full report, plots, raw CSVs, dependency snapshots, and same-commit control](https://github.com/vtjeng/MIPVerify.jl/blob/0232f14ee6cb2983ee83340894f7278739abe289/pairs/2026-07-16-pr222-batched-duals-wk17a-lp-v3/report.md)

## Validation

- `julia --project=test -e 'include("test/net_components/lp_certification.jl")'` — 68/68 passed
- `julia --project=test test/runtests.jl` — 632 passed, 1 expected broken
- JuliaFormatter 1.0.2 — clean
- Directional 2,000-row certificate benchmark — 2.068 ms batched versus 5.141 ms forced scalar
  fallback, with 56% fewer allocated bytes

Closes #211

### CODE REVIEW

**Comment by vtjeng** [comment]
## Performance report: WK17a, LP tightening, 500 samples

PR #222 batches LP certificate dual reads: certificate construction reads each affine constraint group's duals through one vector `ConstraintDual` query instead of one scalar query per row, a performance change that cuts non-solver bound overhead.

This benchmark measures the performance of the verification code. Samples 1–500 of the MNIST test set are verified against the `MNIST.WK17a_linf0.1_authors` network on the master and feature commits under identical settings. The ratio distributions, scatter plots, and outcome-flip tables below compare each sample's candidate run with its own baseline run; the absolute-runtime distributions summarize each side separately. Raw per-sample CSVs, tightening and ReLU rows, dependency snapshots, same-commit control, stats, and plots are in the [`pairs/2026-07-16-pr222-batched-duals-wk17a-lp-v3`](https://github.com/vtjeng/MIPVerify.jl/tree/0232f14ee6cb2983ee83340894f7278739abe289/pairs/2026-07-16-pr222-batched-duals-wk17a-lp-v3) folder at pinned report commit [`0232f14`](https://github.com/vtjeng/MIPVerify.jl/commit/0232f14ee6cb2983ee83340894f7278739abe289).

- **Baseline** post-#209 master `8a455e2756a0d45e224bb1da95f2de8dc2ba3df4`; **candidate** `5b407044078c1ea6108b893f8fd81d8d7f5538c3` (benchmarked implementation; current head `849f535ae971a55b70829774250e7943c54e3b39` contains later cleanup/refactoring and documentation changes and was not re-benchmarked).
- Both commits were benchmarked with `benchmarks/benchmark_wk17a_first100.jl --samples 1:500 --tightening lp --main-time-limit 120 --norm-order Inf` on Julia 1.12.6, single-threaded. All solves (the bound-tightening LPs and the final verification MIP) use HiGHS, an open-source LP/MIP solver. The two runs executed sequentially on a local WSL2 workstation with an identical dependency snapshot (`1cfef4c977ff08219a888aa479cb94eea0b3dbc16f654d1fc0a09167c8f1c74a`); absolute times are not comparable to the CI-hosted `benchmark-results` series.

---

## Summary

- Formulation time fell 18.9%, from 866.2 s to 702.6 s, saving 163.6 s. A 10,000-resample paired input bootstrap gives a 17.6–20.0% interval. This interval measures sensitivity to the fixed input mix; it does not capture machine-to-machine or run-to-run timing variation.
- Non-solver bound overhead (the Non-solver work within bound tightening row) fell 38.6%, from 299.0 s to 183.5 s, saving 115.5 s. This is bound tightening time minus time inside HiGHS. HiGHS bound-solver wall time changed by 0.4 s, and both sides made 99,067 bound-solver calls.
- Summed end-to-end sample time fell 6.6%, from 2,431.7 s to 2,270.7 s, saving 161.0 s across 500 inputs. Whole-run elapsed time changed by the same 161.0 s. The final verification solves were flat in aggregate, so the measured saving came before the final solve.
- The pre-solve improvement was broad: the Build + bound tightening series had a median candidate/baseline ratio of 0.81, with 93% of modeled samples improving by more than 1%.
- Final-solve timing and outcomes varied near the 120 s limit. The candidate resolved two inputs that were unresolved in the baseline, while no semantic outcome regressed. A same-commit control confirms substantial fresh-process HiGHS path variation, so these outcome changes are not used to attribute the formulation speedup.

## Detailed statistics

### Plots

Build + bound tightening moves left; the amount of time for the main solve remains the same. The bound solver calls distributions overlap exactly because both sides made 99,067 calls; the timing gain comes from reducing certificate overhead per solve.

Paired ratio distributions

Absolute runtime distributions

Paired runtime scatter

Absolute bound-call distributions

Paired bound-call scatter

### Per-sample ratio distribution

| series | n | min | p10 | p25 | median | p75 | p90 | max | improved | regressed |
|---|--:|--:|--:|--:|--:|--:|--:|--:|--:|--:|
| Build + bound tightening | 492 | 0.55 | 0.72 | 0.77 | 0.81 | 0.85 | 0.94 | 1.28 | 93% | 6% |
| Main solve time | 492 | 0.03 | 0.79 | 0.91 | 1.01 | 1.12 | 1.28 | 5.19 | 42% | 49% |
| Total end-to-end time | 492 | 0.04 | 0.72 | 0.78 | 0.82 | 0.86 | 0.99 | 4.58 | 90% | 8% |
| Bound solver calls | 492 | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 0% | 0% |

- `ratio` = candidate ÷ baseline; < 1 = candidate faster. `improved` counts samples with ratio below 0.99, `regressed` counts ratio above 1.01; samples within that ±1% band count as unchanged.
- `build` = constructing the MIP model; `tightening` = the LP bound-tightening pass; `main solve` = the final verification MIP; `bound solver calls` = the count of HiGHS bound-tightening solves.
- `total` = `build` + `tightening` + `main solve`.
- The distribution and aggregate tables cover n = 492, the 500 samples minus the 8 `SKIPPED_PREDICTED_IN_TARGETED` samples (see the status table below); those 8 carry no timing and are excluded from all four series.

### Aggregate saving and concentration

| series | baseline | candidate | net saved | pooled ratio | top-10 concentration |
|---|--:|--:|--:|--:|--:|
| Build + bound tightening | 883 s | 719 s | +164 s | 0.81 | 6% |
| Main solve time | 1549 s | 1552 s | −3 s | 1.00 | 91% |
| Total end-to-end time | 2432 s | 2271 s | +161 s | 0.93 | 65% |
| Bound solver calls | 99067 calls | 99067 calls | +0 calls | 1.00 | 0% |

- `net saved` = baseline − candidate total; positive = candidate cheaper.
- `pooled ratio` = candidate total ÷ baseline total (the aggregate counterpart to the per-sample median).
- `top-10 concentration` = the 10 samples with the largest absolute change account for this share of the total absolute per-sample change (0–100%; higher = a few samples dominate). Bound solver calls made 99,067 calls on both sides, so their total absolute per-sample change is 0 and their concentration is reported as 0%.

The 10 largest end-to-end changes account for 65% of the sum of absolute per-input timing changes. Almost all of those 10 changes came from the final solve: 365.5 s of absolute final-solve changes versus 4.9 s of formulation changes. Their gains and losses mostly cancel in aggregate. For build and tightening, the 10 largest changes account for only 6%, so that saving is spread across inputs. A paired input bootstrap for the end-to-end saving spans −5.2% to 19.3%; it is not a system-performance confidence interval.

The component timings below decompose the summed end-to-end time into formulation and final-solve phases. They use the analyzer's dedicated formulation and final-solve timers, while the Build + bound tightening and Main solve time rows above come from a total-minus-solve split, so the subtotals here do not match those two rows one for one; both decompositions cover the full end-to-end time.

#### Component timings

Rows sharing a parent add to that subtotal before rounding.

| component                                     | parent subtotal    |  baseline | candidate |   saved | change |
| --------------------------------------------- | ------------------ | --------: | --------: | ------: | -----: |
| **Summed end-to-end sample time**             | (top level)        | 2,431.7 s | 2,270.7 s | 161.0 s |   6.6% |
| ↳ **Formulation subtotal**                    | Summed sample time |   866.2 s |   702.6 s | 163.6 s |  18.9% |
| ↳ ↳ **Bound-tightening subtotal**             | Formulation        |   469.6 s |   353.8 s | 115.8 s |  24.7% |
| ↳ ↳ ↳ HiGHS bound-solver wall time            | Bound tightening   |   170.6 s |   170.2 s |   0.4 s |   0.2% |
| ↳ ↳ ↳ Non-solver work within bound tightening | Bound tightening   |   299.0 s |   183.5 s | 115.5 s |  38.6% |
| ↳ ↳ Other formulation work                    | Formulation        |   396.6 s |   348.9 s |  47.7 s |  12.0% |
| ↳ Final-solve wall time                       | Summed sample time | 1,562.1 s | 1,564.7 s |  −2.6 s |  −0.2% |
| ↳ Other work inside the sample call           | Summed sample time |    3.34 s |    3.32 s |  0.02 s |   0.7% |

- **Non-solver work within bound tightening:** bound-tightening time minus HiGHS wall time. It includes interval propagation, certificate handling, and bound-loop work for ReLU-scoped bounds.
- **Other formulation work:** formulation time minus bound-tightening time. It includes target and objective construction, optimizer setup, and non-solver work for bounds outside a ReLU layer.
- **Other work inside the sample call:** sample time minus formulation and final-solve wall time.

Each modeled input solves 18 full-network LP bounds for its nine nontarget output logits, producing 8,856 bound solves outside a ReLU layer. LP certificate construction runs after each timed `optimize!` call. For these output-logit bounds, the benchmark assigns only HiGHS wall time to bound tightening; certificate work lands in other formulation work. Their HiGHS time was flat at 54.6 s, while other formulation work fell by 47.7 s. This is consistent with batched dual reads reducing certificate overhead in both the Non-solver work within bound tightening and Other formulation work rows. The benchmark does not time certificate construction separately, so it cannot attribute every second of the residual change directly.

#### Related timing views

These rows use separate or overlapping timing views; do not add them to the table above.

| measurement                             | relationship                                   |  baseline | candidate |   saved | change |
| --------------------------------------- | ---------------------------------------------- | --------: | --------: | ------: | -----: |
| Whole-run elapsed                       | Summed sample time + benchmark-loop overhead   | 2,434.8 s | 2,273.7 s | 161.0 s |   6.6% |
| Formulation excluding bound-solver time | Non-solver work within bound tightening + Other formulation work |   695.6 s |   532.4 s | 163.2 s |  23.5% |

The non-solver bound result addresses issue #211. After progressive tightening, certificate and bound-loop overhead still exceeded time inside HiGHS, and batched reads removed 115.5 s without changing the number of solves.

### Solve status and verdict flips

| status | post-#209 master `8a455e2` | PR #222 `5b40704` |
|---|--:|--:|
| INFEASIBLE | 475 | 476 |
| OPTIMAL | 10 | 10 |
| SKIPPED_PREDICTED_IN_TARGETED | 8 | 8 |
| TIME_LIMIT | 7 | 6 |

Three samples changed solve status; two changed semantic outcome. The number of `time_limit_unresolved` outcomes fell from three to one, and no semantic outcome regressed from a resolved to an unresolved result.

Solve status:

| transition | n | samples |
|---|--:|---|
| `OPTIMAL` → `TIME_LIMIT` | 1 | 150 |
| `TIME_LIMIT` → `INFEASIBLE` | 1 | 407 |
| `TIME_LIMIT` → `OPTIMAL` | 1 | 321 |

Semantic outcome:

| transition | n | samples |
|---|--:|---|
| `time_limit_unresolved` → `adversarial_example_found_or_best_known` | 1 | 212 |
| `time_limit_unresolved` → `certified_no_adversarial_example` | 1 | 407 |

#### Model and outcome audit

The paired raw rows have identical values for:

- dependency snapshot and benchmark arguments;
- variable, binary-variable, structural-constraint, and total-constraint counts;
- all ReLU classification counts and all non-time ReLU-layer fields;
- bound requests, solver calls, statuses, skips, barrier iterations, and nodes.

Bound simplex iterations differed by 834 out of about 2.72 million (0.031%). The two sides ran in fresh Julia and HiGHS processes, and their final MIP search paths also differed.

Of 486 inputs with objective values on both sides, 478 agreed within `1e-6`. The largest differences were timeout-limited incumbents. Sample 291 was the one large optimal/optimal difference: the full baseline reported 0.0461418 and the candidate reported 0.0409316. A same-commit control reran samples 9, 291, and 496 in two fresh processes. Both control runs reported 0.0409316 for sample 291, matching the candidate, and the maximum control objective difference was `5.55e-6`. The control's three final solves also differed by 11.2 s in aggregate with identical code. These observations are consistent with fresh-process solver variation and are not consistent with a semantic regression from batched reads.
